plugins/prometheus/prometheus.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2017 Mesosphere, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+	plugin "github.com/dcos/dcos-metrics/plugins"
+	"github.com/dcos/dcos-metrics/producers"
+	"github.com/urfave/cli"
+)
+
+var (
+	pluginFlags = []cli.Flag{
+		cli.IntFlag{
+			Name:   "prometheus-port",
+			Usage:  "The port on which to serve prometheus metrics",
+			EnvVar: "PROMETHEUS_PORT",
+			Value:  8080,
+		},
+	}
+	registerOnce sync.Once
+	listener     net.Listener
+)
+
+func main() {
+	log.Info("Starting statsd DC/OS metrics plugin")
+
+	promPlugin, err := plugin.New(
+		plugin.Name("prometheus"),
+		plugin.ExtraFlags(pluginFlags),
+		plugin.ConnectorFunc(promConnector))
+
+	promPlugin.BeforeFunc = startPromServer
+	promPlugin.AfterFunc = stopPromServer
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Fatal(promPlugin.StartPlugin())
+}
+
+func startPromServer(c *cli.Context) error {
+	addr := fmt.Sprintf(":%d", c.Int("prometheus-port"))
+	log.Infof("Starting Prometheus server on localhost%s", addr)
+
+	registerOnce.Do(func() {
+		// the server may start and stop, but the handler must only be
+		// registered once
+		http.HandleFunc("/metrics", serveMetrics)
```

Comment:
I think if you create an http.Server instance each time the prom server starts, you can register a handler for a path on that server instance instead of needing to set it globally here.

Reply:
The plugin starts and stops the server only once in its lifecycle, but the tests start it repeatedly. Unless you're strongly opposed, I'd prefer to leave it as it is.